sopel/irc/modes.py

```diff
@@ -0,0 +1,234 @@
+        """Map of mode types (``str``) with their list of modes (``tuple``).
+
+        This map should come from ``ISUPPORT``, usually through
+        :attr:`bot.isupport.CHANMODES <sopel.irc.isupport.ISupport.CHANMODES>`.
+        """
+        self.type_params = type_params
+        """Map of mode types (``str``) with their param requirements.
+
+        This map default to :data:`DEFAULT_MODETYPE_PARAM_CONFIG`.
+        """
+        self.privileges = privileges
+        """Set of valid user privileges.
+
+        This set should come from ``ISUPPORT``, usually through
+        :attr:`bot.isupport.PREFIX <sopel.irc.isupport.ISupport.PREFIX>`.
+
+        If a server doesn't advertise its prefixes for user privileges,
+        :attr:`PRIVILEGES` can be used as a default value.
+        """
+
+    def get_mode_type(self, mode: str) -> str:
+        """Retrieve the type of ``mode``.
+
+        :raise ModeTypeUnknown: if the mode's type cannont be determined
+        :return: the mode's type as defined by :attr:`chanmodes`
+
+        ::
+
+            >>> mm = ModeMessage({'A': tuple('beI'), 'B': tuple('k')}, {})
+            >>> mm.get_mode_type('b')
+            'A'
+            >>> mm.get_mode_type('k')
+            'B'
+
+        This method will raise a :exc:`ModeTypeUnknown` if the mode is unknown,
+        including the case where ``mode`` is actually a user privilege.
+        (such as ``v``)
+        """
+        for letter, modes in self.chanmodes.items():
+            if mode in modes:
+                return letter
+        raise ModeTypeUnknown(mode)
+
+    def get_mode_info(
+        self,
+        mode: str,
+        is_added: bool
+    ) -> Tuple[Optional[str], bool, bool]:
+        """Retrieve ``mode``'s information when added or removed.
+
+        :raise ModeTypeUnknown: when the mode's type is unknown and isn't a
+                                user privilege
+        :raise ModeTypeImproperlyConfigured: when the mode's type is known but
+                                             there is no information for
+                                             parameters (if and when they are
+                                             required by the mode)
+        :return: a tuple with three values: the mode type, if it requires a
+                 parameter, and if it's a channel mode or a user privilege
+
+        ::
+
+            >>> chanmodes = {'A': tuple('beI'), 'B': tuple('k')}
+            >>> t_params = {'A': PARAM_ALWAYS, 'B': PARAM_ADDED}
+            >>> mm = ModeMessage(chanmodes, t_params)
+            >>> mm.get_mode_info('e', False)
+            ('A', True, False)
+            >>> mm.get_mode_info('k', False)
+            ('B', False, False)
+            >>> mm.get_mode_info('v', True)
+            (None, True, True)
+
+        .. note::
+
+            A user privilege ``mode`` doesn't have a type so the first value
+            returned will be ``None`` in that case.
```

Comment:
This is as good a place to ask as any: Is this because user privilege modes are given in `PREFIX` instead of `CHANMODES`, so they're not technically assigned to a category?

I wonder if `None` is really the best way to indicate such a thing, instead of another more meaningful value.

Reply:
Eeyup, that's why: it doesn't have a type. I use `None` for that, because that's what makes sense to me.